### PR TITLE
Feign propagated URI templates for tracing

### PR DESCRIPTION
**Pull request by markelliot**

### CODE REVIEW

**Comment by uschi2000** [comment]
Yep, this is the hack I had discussed with James, too. Mark, are you sending this is a proof of concept, or do you you think we should implement this? In the latter case, also need Retrofit.

**Comment by markelliot** [comment]
Had some time this morning and implemented the hack you proposed.

**Comment by markelliot** [comment]
(so that we could merge)

**Comment by markelliot** [comment]
Retrofit doesn't have a great way to intercept the call before it's passed to OkHttp, so what I'd propose instead of including the solution here, is that we use the `@Headers` annotation to where we generate Retrofit interfaces such that the path is added as a header in the 'normal' retrofit scheme. The OkHttp client change in this PR will result in that, then, safely being included in the trace logs.

Cool?

**Comment by uschi2000** [comment]
Uh, that's quite convoluted :( Is that really worth the effort, given that we have the API endpoint name in the corresponding server span already?

**Comment by markelliot** [comment]
The vast majority of the callers we use are Feign-based, even with Retrofit this is a strict improvement for those callers and no worse for Retrofit. If the Retrofit workaround is too gross, perhaps we can take this just as upside on Feign, for now.

**Comment by markelliot** [comment]
...am still digging around Retrofit, think maybe there's an option to fix this

**Comment by markelliot** [comment]
...j/k, there's not really a good pre-request Retrofit interceptor, so we'd have to duplicate and slightly tweak a fairly chunky bit of code.

If you don't want to take the Feign upside alone, you can close the PR.

**Comment by uschi2000** [comment]
OK, let's do it. Want to add a simple test?

File: tracing-okhttp3/src/main/java/com/palantir/remoting3/tracing/okhttp3/OkhttpTraceInterceptor.java [hunk 1]
```diff
@@ -29,10 +29,20 @@
 public enum OkhttpTraceInterceptor implements Interceptor {
     INSTANCE;
 
+    private static final String HTTP_REMOTING_HEADER = "hr-path-template";
```
Comment (uschi2000) [inline tracing-okhttp3/src/main/java/com/palantir/remoting3/tracing/okhttp3/OkhttpTraceInterceptor.java]: let's call this PATH_TEMPLATE_HEADER, and reuse in the contract

Comment (uschi2000) [reply tracing-okhttp3/src/main/java/com/palantir/remoting3/tracing/okhttp3/OkhttpTraceInterceptor.java]: (not sure you can reuse...)

**Comment by markelliot** [comment]
ok, all set.

**Comment by markelliot** [comment]
something with GH/circle seems messed up, though

**Review by uschi2000 — APPROVED** [review]

**Comment by j-baker** [comment]
Depressingly this does not work. Basically because the template URL is in a template header, Feign just templates everything into it.

**Pull request merged**